File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
def solution(name):
    answer = 0
    l = len(name)
    
    # 각 알파벳 시작은 A
    # 1. 해당 알파벳 위 아래 움직임 계산
    for n in name :
        answer += min(ord("Z") - ord(n) + 1, ord(n) - ord("A"))
    
    move = l - 1
    for i in range(l) :
        # 2. 좌우 움직임 고려하기
        n = i + 1
        while n < l and name[n] == "A" : n += 1 
        
        move = min(move, i + i + (l - n), (l - n) * 2 + i )
        
    return answer + move

File: 프로그래머스/2/42860. 조이스틱/조이스틱.java
class Solution {
    public int solution(String name) {
        int answer = 0;
        
        // 글자 위아래로 움직이기
        for(char c : name.toCharArray()){
            int zGap = 'Z' - c + 1;
            int aGap = c - 'A';
            
            if(zGap >= aGap){answer += aGap;}
            else{answer+= zGap;}
        }
        
        int len = name.length();
        int move = len - 1;
        for(int i = 0; i < len; i++){
            int next = i + 1;
            while(next < len && name.charAt(next) == 'A'){
                next ++;
            }
            
            move = Math.min(move, i * 2 + (len - next));
            move = Math.min(move, i + (len - next) * 2);
        }
        
        return answer + move;
    }
}